strawberryfields/io/xir_io.py

```diff
@@ -0,0 +1,294 @@
+                    gate() | regrefs  # pylint:disable=expression-not-assigned,pointless-statement
+                else:
+                    gate | regrefs  # pylint:disable=expression-not-assigned,pointless-statement
+
+    prog._target = xir_prog.options.get("target", None)  # pylint: disable=protected-access
+
+    if "shots" in xir_prog.options:
+        prog.run_options["shots"] = xir_prog.options["shots"]
+    if "cutoff_dim" in xir_prog.options:
+        prog.backend_options["cutoff_dim"] = xir_prog.options["cutoff_dim"]
+
+    return prog
+
+
+def to_xir(prog, **kwargs):
+    """Convert a Strawberry Fields Program to an XIR Program.
+
+    Args:
+        prog (Program): the Strawberry Fields program
+
+    Keyword Args:
+        add_decl (bool): Whether gate and output declarations should be added to
+            the XIR program. Default is ``False``.
+        version (str): Version number for the program. Default is ``0.1.0``.
+
+    Returns:
+        xir.Program
+    """
+    version = kwargs.get("version", "0.1.0")
+    xir_prog = xir.Program(version=version)
+
+    if isinstance(prog, TDMProgram):
+        xir_prog.add_option("type", "tdm")
+        xir_prog.add_option("N", prog.N)
+        for i, p in enumerate(prog.tdm_params):
+            xir_prog.add_constant(f"p{i}", _listr(p))
+
+    if prog._target:
+        xir_prog.add_option("target", prog._target)  # pylint: disable=protected-access
+    if "cutoff_dim" in prog.run_options:
+        xir_prog.add_option("cutoff_dim", prog.run_options["cutoff_dim"])
+    if "name" in prog.run_options:
+        xir_prog.add_option("name", prog.run_options["name"])
+    if "shots" in prog.run_options:
+        xir_prog.add_option("shots", prog.run_options["shots"])
+
+    # fill in the quantum circuit
+    for cmd in prog.circuit or []:
+
+        name = cmd.op.__class__.__name__
+        wires = tuple(i.ind for i in cmd.reg)
+
+        if "Measure" in name:
+            if kwargs.get("add_decl", False):
+                output_decl = xir.Declaration(name, type_="out")
+                xir_prog.add_declaration(output_decl)
+
+            params = {}
+            if cmd.op.p:
+                # argument is quadrature phase
+                a = cmd.op.p[0]
+                if a in getattr(prog, "loop_vars", ()):
+                    params["phi"] = a.name
+                else:
+                    params["phi"] = a
+
+            # special case to take into account 'select' keyword argument
+            if cmd.op.select is not None:
+                params["select"] = cmd.op.select
+
+            if name == "MeasureFock":
+                # special case to take into account 'dark_counts' keyword argument
+                if cmd.op.dark_counts is not None:
+                    params["dark_counts"] = cmd.op.dark_counts
+        else:
+            if kwargs.get("add_decl", False):
+                if name not in [gdecl.name for gdecl in xir_prog.declarations["gate"]]:
```

Comment:
I wonder if we can use `xir.Program.search()` here.

Reply:
No, that's not possible. `xir.Program.search` only returns the parameters or wires of a declaration from a name and type. We're looking to get a list of all gate declaration names.

Reply:
Oh, I was thinking of something along the lines of
```python
try:
    program.search("gate", "params", name)
except ValueError:
    # Not found.
else:
    # Found.
```
This is probably not a good idea though!

Reply:
Oh, I see. I guess that'd be another solution. 🤔